File: 142.环形链表-ii.py
#
# @lc app=leetcode.cn id=142 lang=python3
#
# [142] 环形链表 II
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: ListNode) -> ListNode:
        """
        # set
        nodeSet = set()
        while head:
            nodeSet.add(head)
            head = head.next
            if nodeSet.__contains__(head):
                return head
        return None
        """
        # double pointer 
        fast, slow = head, head
        while True:
            if fast == None or fast.next == None:
                return None
            fast = fast.next.next
            slow = slow.next
            if slow == fast:
                break
        
        slow = head
        while slow != fast:
            fast = fast.next
            slow = slow.next
        return slow

# @lc code=end

File: 142.环形链表-ii.java
import java.util.HashSet;
import java.util.Set;

/*
 * @lc app=leetcode.cn id=142 lang=java
 *
 * [142] 环形链表 II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        /*hashset
        Set<ListNode> nodeSet = new HashSet<>();
        while (head != null) {
            nodeSet.add(head);
            head = head.next;
            if (nodeSet.contains(head)) {
                return head;
            }
        }
        return null;
        */
        // double pointer
        ListNode fast, slow;
        fast = slow = head;
        while (true) {
            if (fast == null || fast.next == null) {
                return null;
            }
            fast = fast.next.next;
            slow = slow.next;
            if (slow == fast) {
                break;
            }
        }
        slow = head;
        while (slow != fast) {
            slow = slow.next;
            fast = fast.next;
        }
        return slow;
    }
}
// @lc code=end

